compiler/circle-mpqsolver/src/utils/ErrorApproximator.cpp

```diff
@@ -0,0 +1,334 @@
+    return;
+  }
+
+  for (indices[0] = 0; indices[0] < dimension.dim(0).value(); indices[0]++)
+  {
+    for (indices[1] = 0; indices[1] < dimension.dim(1).value(); indices[1]++)
+    {
+      for (indices[2] = 0; indices[2] < dimension.dim(2).value(); indices[2]++)
+      {
+        for (indices[3] = 0; indices[3] < dimension.dim(3).value(); indices[3]++)
+        {
+          func(indices, dimension, channel_dim_index);
+        }
+      }
+    }
+  }
+}
+
+static void cal_minmax_per_channel(CircleConst *node, std::vector<float> &min,
+                                   std::vector<float> &max, int32_t &channel_dim_index)
+{
+  loco::TensorShape dimension;
+  dimension.rank(4);
+
+  if (!get_channel_dim_index(node, dimension, channel_dim_index))
+  {
+    throw std::runtime_error("Failed to find channel index in " + node->name());
+  }
+  auto size = dimension.dim(channel_dim_index).value();
+
+  std::vector<bool> has_min_max_value(size, false);
+  min.resize(size);
+  max.resize(size);
+
+  auto cal_minmax = [&](uint32_t *indices, loco::TensorShape &dimension, int channel_dim_index) {
+    int channel_idx = indices[channel_dim_index];
+    auto data = node->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices));
+    if (has_min_max_value[channel_idx])
+    {
+      min[channel_idx] = data < min[channel_idx] ? data : min[channel_idx];
+      max[channel_idx] = data > max[channel_idx] ? data : max[channel_idx];
+    }
+    else
+    {
+      min[channel_idx] = data;
+      max[channel_idx] = data;
+      has_min_max_value[channel_idx] = true;
+    }
+  };
+
+  iterate_per_channel(node, channel_dim_index, cal_minmax);
+}
+
+bool getShape(CircleNode *circle_node, std::vector<uint32_t> &shape)
+{
+  if (circle_node->shape_status() == ShapeStatus::VALID)
+  {
+    auto rank = circle_node->rank();
+    if (rank != 4)
+      return false;
+
+    shape.resize(rank);
+    for (int i = 0; i < rank; i++)
+    {
+      shape[i] = circle_node->dim(i).value();
+    }
+    return true;
+  }
+
+  return false;
+}
+
+static void get_values_per_channel(CircleConst *node, std::vector<std::vector<float>> &values,
+                                   int32_t &channel_dim_index)
+{
+  loco::TensorShape dimension;
+  dimension.rank(4);
+
+  if (!get_channel_dim_index(node, dimension, channel_dim_index))
+  {
+    throw std::runtime_error("Failed to find channel index in " + node->name());
+  }
+  auto size = dimension.dim(channel_dim_index).value();
+
+  std::vector<bool> has_min_max_value(size, false);
+  values.resize(size);
+
+  auto cal_minmax = [&](uint32_t *indices, loco::TensorShape &dimension, int channel_dim_index) {
+    int channel_idx = indices[channel_dim_index];
+    auto data = node->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices));
+    values[channel_idx].push_back(data);
+  };
+
+  iterate_per_channel(node, channel_dim_index, cal_minmax);
+}
+} // namespace
+
+float ErrorApproximator::Approximate(CircleNode *node)
```

Comment:
> Can you explain what this function does?

It tries to estimate maximal layer error introduced by quantization, to recognize where the most sensitive to quantization layers are (closer to input nodes or to output nodes).

>It would be better to add comments.

Ok.

Reply:
> layer error

Detailed equation to compute the layer error would be helpful. For example, mse or mae, etc.

Reply:
Well, It seems that the error is aggregate of each layer's upper bound of L1 quantization error. Am I right?

In the non-draft code, the equation will be included in the comments.

Reply:
> Well, It seems that the error is aggregate of each layer's upper bound of L1 quantization error. Am I right?

Yes. You are.

>In the non-draft code, the equation will be included in the comments.

Ok.

Reply:
Let's be more precise for estimation of convolutions. Please correct me, if i am wrong.
![image](https://user-images.githubusercontent.com/112689352/211476430-cac471a5-72b1-4d33-9dca-9e53fe45ed0e.png)
[convolution.pdf](https://github.com/Samsung/ONE/files/10379502/convolution.pdf)